### PR TITLE
Validate endpoints from feature flags

### PR DESCRIPTION
**Pull request by ato9000**
[Preview this branch](https://cosmos-explorer-preview.azurewebsites.net/pull/1196)

### CODE REVIEW

File: src/Utils/EndpointValidation.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    `^https:\\/\\/[\\.\\w]*portal\\.azure\\.(com|cn|us)$`,
+    `^https:\\/\\/[\\.\\w]*portal\\.microsoftazure.de$`,
+    `^https:\\/\\/[\\.\\w]*ext\\.azure\\.(com|cn|us)$`,
+    `^https:\\/\\/[\\.\\w]*\\.ext\\.microsoftazure\\.de$`,
+    `^https://cosmos-db-dataexplorer-germanycentral.azurewebsites.de$`,
+];
+
+export const allowedJunoOrigins: ReadonlyArray<string> = [
+    "https://juno-test.documents-dev.windows-int.net",
+    "https://juno-test2.documents-dev.windows-int.net",
```
Comment (gitanuj) [inline src/Utils/EndpointValidation.ts:31]: Endpoint for juno test3 is missing. It should include `https://juno-test3.documents-dev.windows-int.net`

Comment (gitanuj) [reply src/Utils/EndpointValidation.ts]: There is a JunoEndpoints constants file from where we can use these endpoints too.

File: src/Utils/EndpointValidation.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+export function validateEndpoint(endpointToValidate: string | undefined, allowedEndpoints: string[]): boolean {
+  if (!endpointToValidate) {
+    return true;
+  }
+  const originToValidate: string = new URL(endpointToValidate).origin;
+  const allowedOrigins: string[] = allowedEndpoints.map((allowedEndpoint) => new URL(allowedEndpoint).origin) || [];
```
Comment (languy) [inline src/Utils/EndpointValidation.ts:6]: Consider optimizing: if `allowedEndpoint` is static, the origin could be computed once. (as in: `const allowedOrigins = ...`).
This said, it's probably not going to matter much either way.

**Review by languy — DISMISSED** [review]

**Review by asierisayas — APPROVED** [review]

**Review by gitanuj — DISMISSED** [review]

File: src/Utils/EndpointValidation.ts (hunk 2 above)
Comment (ato9000) [reply src/Utils/EndpointValidation.ts:19]: Let's discuss offline. I will leave as is for now.

**Review by asierisayas — DISMISSED** [review]

File: src/Utils/EndpointValidation.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import * as Logger from "../Common/Logger";
+
+export function validateEndpoint(endpointToValidate: string | undefined, allowedEndpoints: string[]): boolean {
```
Comment (victor-meng) [inline src/Utils/EndpointValidation.ts:3]: Can the `allowedEndpoints` be `ReadyonlyArray<string> | string[]` so we don't have to convert the ReadonlyArrays to a regular array?

File: src/Explorer/Explorer.tsx [hunk 4]
```diff
@@ -422,7 +417,13 @@ export default class Explorer {
     connectionStatus.status = ConnectionStatusType.Connected;
     useNotebook.getState().setConnectionInfo(connectionStatus);
     useNotebook.getState().setNotebookServerInfo({
-      notebookServerEndpoint: userContext.features.notebookServerUrl || connectionInfo.data.notebookServerUrl,
+      notebookServerEndpoint:
+        (validateEndpoint(
+          userContext.features.notebookServerUrl,
+          allowedNotebookServerUrls.map((endpoint) => endpoint)
```
Comment (victor-meng) [inline src/Explorer/Explorer.tsx:423]: Is there anyway we can avoid doing this conversion? Also we can use the spread operator to do this if we have to: `[...allowedNotebookServerUrls]`

**Review by victor-meng — APPROVED** [review]

**Review by gitanuj — APPROVED** [review]

**Review by victor-meng — DISMISSED** [review]

**Pull request merged**